lib.attrsets: Add genAttrs'

cyntheticfox opened the pull request:
Add a relatively simple function to allow creating attrsets from lists with configurable key names.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Create an additional library function I personally use often that I believe others could benefit from.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

I wasn't 100% sure how to order the two function arguments. In most cases, I see the primary parameter as the last parameter so as to allow for eta reductions, as with `mapAttrs'`, but `genAttrs` has the "names" list as the first parameter for some reason.

I also wasn't sure if I should add `map` as an inherit alongside `head`, `tail`, and `length` on line 5, since some functions are there but others aren't.

cyntheticfox committed:
lib.attrsets: Add genAttrs'

Add a relatively simple function to allow creating attrsets from lists with
configurable key names.

Revised to have example output order reflect lexigraphical ordering of
attrset keys.

alyssais reviewed (commented):
I do think this is functionality that would be helpful, but I think it would be better called `mapListToAttrs` for consistency with `listToAttrs`, `mapAttrsToList`, etc.

houstdav000 commented:
I always did find "`genAttrs`" to be a rather vague name... I think `mapListToAttrs` would be much clearer.

Would it then be best to rename the existing `genAttrs` function to such and this additional function to `mapListToAttrs'`, since this is a variant of the former?

cyntheticfox committed:
lib.attrsets: Deprecate genAttrs with rename

Create a more descriptive function name `mapListToAttrs` for `genAttrs`
as well as flip the arguments around to allow for eta reduction. Also
rename the corresponding prime function.

Additionally, correct the `genAttrs` example output to be
lexigraphically sorted as AttrSet outputs are.

houstdav000 commented:
Renamed to `mapListToAttrs` and `mapListToAttrs'` while providing deprecation warning for `genAttrs` in 16cda89.

roberth commented on lib/attrsets.nix:
```diff
@@ -534,23 +534,42 @@ rec {
+    list:
+    listToAttrs (map (n: nameValuePair n (f n)) list);
+
+  # DEPRECATED
+  genAttrs = lib.warn
+    "lib.genAttrs is deprecated, use lib.mapListToAttrs instead" (flip mapListToAttrs);
```
This is an old, widespread function.
Deprecating `genAttrs` now would cause third-party library maintainers to migrate and therefore break with previous nixpkgs versions, among which is the 22.11 installer.
I doubt that it should be deprecated at all because it is widespread, and because having the values and the variable name close together is nice.

houstdav000 replied:
It might indeed be good to deprecate it at a later time, such as with a release, and leave it as an action item commented as such for later. 

Not deprecating _ever_ because it's used doesn't make much sense to me though, as the name isn't indicative of its purpose in typical functional programming terms, and the parameter ordering as-is honestly doesn't make sense -- it's not consistent with the rest of the `lib.attrsets` functions, and afaik, it's rather standard to have the applied data after any function parameters for higher-order functions in reflection of a typed lambda calculus

houstdav000 replied:
Removed warning in 6f0562b. Leaving comment unresolved for now.

roberth commented on lib/attrsets.nix:
```diff
@@ -534,23 +534,42 @@ rec {
 
      Type:
-       genAttrs :: [ String ] -> (String -> Any) -> AttrSet
+       mapListToAttrs :: [String] -> (String -> Any) -> AttrSet
```
None of this documentation takes into account the flipped order.

houstdav000 replied:
Fixed in 195c0a4

roberth reviewed (changes requested):
Please restore `genAttrs`, and feel free to add the new function.

cyntheticfox committed:
lib/attrsets: Make genAttrs deprecation "TODO"

As roberth pointed out, deprecating mid release-cycle would cause
unintended breakage for both many packages and with the 22.11
installer, making use of 22.11 install ISOs to install unstable
difficult.

I left the warning code commented-out for ease of later use.

cyntheticfox committed:
lib/attrsets: Fix mapListToAttrs arg flip

As roberth pointed out, the documentation of the `mapListToAttrs`
function, renamed from `genAttrs` and having its arguments flipped, did
not have the changes in argument ordering reflected in the
documentation.

Updating the documentation example and type to reflect the modification.

cyntheticfox committed:
lib/attrsets: Fix genAttrs tests

Change names and orderings such that tests should now pass.

infinisil commented on lib/attrsets.nix:
```diff
@@ -534,23 +534,53 @@ rec {
+     attribute names.
+
+     Example:
+       mapListToAttrs (name: "x_" + name) [ "foo" "bar" ]
+       => { bar = "x_bar"; foo = "x_foo"; }
```
`mapAttrsToList` should expectedly be the inverse of `mapListToAttrs`, but that's not the case here

roberth replied:
> Should `mapAttrsToList` only work on values instead?

`mapAttrsToList` is a ten year old function that works well. Changing it will do more harm than good.

A better framework for understanding the function is that it's `map f ∘ attrsToList`.

> With the preexisting built-in `mapAttrs` function working on only the values, we had to add a `mapAttrs'`.

> I personally believe either `builtins.mapAttrs` should effectively be `lib.mapAttrs'`,

The apostrophe generally refers to the function being either new, or more strict. Both apply in this case. `mapAttrs` can return the attribute set spine (`attrNames` + thunks) without evaluating any of the `attrValues`, whereas `mapAttrs'` is generally more strict, having to evaluate the function for each of the attributes. It does depend on the strictness of the passed function.
Defining a "map" function to operate on single values is quite conventional. It is originally a transformation of a function, although "multi-paradigm" languages may have you believe otherwise; in Nix it is a transformation of a function, and `mapAttrs` is more general, allowing any function to operate on attribute sets instead. `mapAttrs'` requires a curried function.

----------

We are well past the stage where we could have made such breaking changes to `lib`, let alone `builtins`. I think we're over ten years late for that.

A redesign would have to be done as `lib2` and would take a lot more than one change to make it worthwhile. Even then it has a huge cost for the ecosystem; an effort that would be better spent on something more productive.

The only reasonable thing we can do here is one or two functions that are consistent with the current patterns, which gets us back to the inverse of

    map f ∘ attrsToList

which would be

    listToAttrs ∘ map g

or equivalently

    mapAttrs h ∘ listToAttrs

In practice, the list items may not already be in the form of name-value pair attrsets, so the `listToAttrs ∘ map g` variant seems easier to use.
If the list items _are_ in a name-value pair form, then it's likely that another function such as `concatMapAttrs` could have done the job instead, so I'm inclined to dismiss that use case.

-------

For context, consistency in a library like this is very important, because remembering the signatures of individual functions is not feasible. Most of these functions do not pass the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold) by themselves and are only a net positive because they form a consistent "language" of their own. This "language" only has somewhat of a right to exist for two reasons, which are
 - performance optimizations
 - lack of a truly convenient syntax for composing long chains of functions

houstdav000 replied:
> The apostrophe generally refers to the function being either new, or more strict. Both apply in this case. mapAttrs can return the attribute set spine (attrNames + thunks) without evaluating any of the attrValues, whereas mapAttrs' is generally more strict, having to evaluate the function for each of the attributes. It does depend on the strictness of the passed function.

Okay, so by "strict", you mean in evaluation steps and in what it produces? I suppose that yes, it is more "general" in the sense that _all_ functions compliant with its type will work with attribute sets, though when I think "generality", I'm thinking of the data on which I'm operating, and what I can do with it, which I suppose would be "flexibility" instead, which mapAttrs' does have.

---

Similarly, I don't see how `listToAttrs ∘ map g` is equivalent to `mapAttrs h ∘ listToAttrs`. `listToAttrs` is predicated on the applied data set being pairs of name-value form, with `concatMapAttrs` operating on one existing attrset. I continually came across the issue of wanting to transform both name and value for lists, and kept running into not having an easy way of expressing it available.

I originally thought to take two separate functions and have it be of type `(A -> String) -> (A -> Any) -> [A] -> AttrSet`, but was pointed out it could be reduced, though IMO at the cost of immediate observable usability.

---

> For context, consistency in a library like this is very important...

Therein is my point. Consistency in _what_? Consistency in the strictness of evaluation? Consistency in the applicable data? Consistency in parameter ordering?

It's a tradeoff between simplicity and usability, no? You can have a library designed for the purpose of a strict language for package writing or testing, but that doesn't explain why the library functions available go beyond such. Isn't it for the developer to not only be able to shorten their code, but to be able to express their intent in different manners and conceptualize their mental models better?

roberth replied:
> Okay, so by "strict", you mean in evaluation steps and in what it produces?

Yes, strictness as in evaluation. A result is strict in an in argument when that argument must be evaluated in order produce the result.

-----

> Similarly, I don't see how `listToAttrs ∘ map g` is equivalent to `mapAttrs h ∘ listToAttrs`.

Equivalent in what the design permits to be expressed. The assumption here is that `g` and `h` are equivalent under some conversion, but I figured that was implied. I should've picked `f` and `f'` perhaps.

-----

> It's a tradeoff between simplicity and usability, no?

It's not; see the Fairbairn threshold link. In general, a single function is not easier to use than a composition of functions, unless the composition is non-trivial.

You may have noticed that any grouping of container-related functions can be refactored by abstracting the domain-specific code into parameters and giving the remainder a name. The Fairbairn threshold comes from roughly the observation that we'll eventually have as many instances of this glue code that the best naming scheme is to eventually just name all the functions after their implementations. From those implementations, only the functions with "significant" functionality need to remain and be part of the naming scheme. So instead of the user composing those significant functions, is it simpler or more usable to introduce a naming scheme that just repeats those functions? And that's assuming you get the naming scheme right and that's it's accurate and memorable. Actual code is already accurate and does not need to be memorable.

As I explained, in Nix we have other reasons to add such functions, which are a bit questionable from a usability and simplicity perspective; they're about keystrokes and - even more questionably - performance.

> Isn't it for the developer to not only be able to shorten their code

Indeed it's not. Shortening must not be the goal, but a positive side effect of other choices. Readability _is_ a goal, and shortening may or may not follow from that.

> but to be able to express their intent in different manners and conceptualize their mental models better?

For domain-specific code, yes, but this is not domain-specific code. `mapListToAttrs` conceptually does not abstract anything from "map + `listToAttrs`", except the order of the two, which is an arbitrary decision anyway. It being arbitrary indicates that it is not domain-specific and will lead to surprises because a user may expect the other decision. This is where consistency comes in.

-----

> Consistency in _what_?

Unwritten rules of thumb that explain why certain functions are named the way they are and how their inputs and outputs relate to similarly named functions. I wish I had a better answer, but that's how it is.

> Consistency in the strictness of evaluation?

As lazy as possible. This is only tangentially relevant here.

> Consistency in the applicable data?

If you'd optimize for this, you'd probably divide the library into sublibraries that operate largely on single types, and you'd reduce the functions that operate between the types to a minimum. This implies removing `mapAttrsToList`, but that ship has sailed.

> Consistency in parameter ordering?

Functions called "map" always take a function as their argument. ("first argument" if curried)

houstdav000 replied:
I can see the reasoning in your replies @roberth, though I can't say I agree with the underlying assumption that developer usability and approachability is intertwined with experienced readability, or that there is one objectively sufficient one-size-fits-all implementation. I do see though that this is getting a little out-of-hand, and isn't as compatible with the existing nixpkgs library as I thought, so I might either simply add it to mine or factor out into my own library repository.

If this sounds amenable, I'll close the pull request or you can, at your discretion.
